ledger-service/jwt/src/main/scala/com/digitalasset/jwt/KeyUtils.scala

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.jwt
+
+import java.io.{File, FileInputStream}
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
+import java.security.cert.CertificateFactory
+import java.security.interfaces.{RSAPrivateKey, RSAPublicKey}
+import java.security.spec.PKCS8EncodedKeySpec
+import java.security.KeyFactory
+
+import com.digitalasset.daml.lf.data.TryOps.Bracket.bracket
+import scalaz.Show
+import scalaz.syntax.show._
+
+import scala.util.Try
+
+object KeyUtils {
+  final case class Error(what: Symbol, message: String)
+
+  object Error {
+    implicit val showInstance: Show[Error] =
+      Show.shows(e => s"PemUtils.Error: ${e.what}, ${e.message}")
+  }
+
+  private val mimeCharSet = StandardCharsets.ISO_8859_1
+
+  /**
+    * Reads an RSA public key from a X509 encoded file.
+    * These usually have the .crt file extension.
+    */
+  def readRSAPublicKeyFromCrt(file: File): Try[RSAPublicKey] = {
+    bracket(Try(new FileInputStream(file)))(is => Try(is.close())).flatMap { istream =>
+      Try(
+        CertificateFactory
+          .getInstance("X.509")
+          .generateCertificate(istream)
+          .getPublicKey
+          .asInstanceOf[RSAPublicKey])
+    }
+  }
+
+  /**
+    * Reads a RSA private key from a PEM/PKCS#8 file.
+    * These usually have the .pem file extension.
+    */
+  def readRSAPrivateKeyFromPem(file: File): Try[RSAPrivateKey] = {
+    bracket(Try(new FileInputStream(file)))(is => Try(is.close())).flatMap { istream =>
+      for {
+        fileContent <- Try(Files.readAllBytes(file.toPath))
+
+        // Remove PEM container header and footer
+        pemContent <- Try(
+          new String(fileContent, mimeCharSet)
+            .replaceFirst("-----BEGIN ([A-Z ])*-----\n", "")
+            .replaceFirst("\n-----END ([A-Z ])*-----\n", "")
+            .replace("\r", "")
+            .replace("\n", "")
+        )
+
+        // Base64-decode the PEM container content
+        decoded <- Base64
+          .decode(pemContent)
+          .leftMap(e => new RuntimeException(e.shows))
+          .toEither
+          .toTry
+
+        // Interpret the container content as PKCS#8
+        key <- Try {
+          val kf = KeyFactory.getInstance("RSA")
```

Comment:
Not sure how often this method is called. If often, we should cache the `KeyFactory` instance? `getInstnace` creates a new object every time it is called.

Reply:
It's called whenever an RSA key needs to be loaded from a file, which is usually once per application life time.